rfcs/2021-09-14-3910-schema-support.md

```diff
@@ -0,0 +1,472 @@
+
+1. For each **source**, the implicit or explicit schemas defined for that source
+   are tracked. Any source that doesn't have a schema defined is considered to
+   expose "any" schema.
+
+2. Each downstream **transform** passes along the merged set of the schema
+   definitions of their upstream components to their child components. That is,
+   if transform `C` receives events from sources `A` and `B`, and `A` exports
+   events adhering to schema `1` and `2` and `B` for schema `2` and `3`, then
+   transform `C` defines itself as exporting events adhering to schema `1`, `2`
+   and `3`.
+
+3. Transforms _can_ overwrite this default behavior. For example, the `route`
+   transform could route events based on schema `1` differently than `2`. Given
+   that the route transform compiles down to one transform per lane, each lane
+   will have a different schema output defined.
+
+   Similarly, the `remap` transform defines its own output schema(s), based on
+   information from the VRL compiler.
+
+4. Each **sink** has an implicit or explicit export schema defined. It too
+   creates a merged set of its input components similar to transforms, but
+   instead of producing a list of exported schemas, it checks that the input
+   schemas can be automatically converted to the required export schema as
+   defined in the sink.
+
+5. If anything in the above chain is invalid, Vector returns an error, and won't
+   boot.
+
+By doing this, We've covered (_almost, see below_) all cases related to schema
+validation, conversions and routing, except for the initial validation happening
+when events are received by a source.
+
+- [ ] **TODO** Give examples on how we're going to expand the different
+      component traits.
+
+#### Run-time validation
+
+This is where run-time validation is required.
+
+Because sources are the _only_ type of component that can ingest events from
+external places, it is also the only place that requires us to do run-time
+validation. All other invariants have already been covered by the boot-time
+validations.
+
+Run-time schema validations are expensive, but there's no way to avoid them. For
+the first implementation of this RFC, we're going to naively check each
+individual event using an existing Rust-based JSON-schema library. In the
+future, we potentially want to look into combining event decoding and schema
+validation into a single step, but this will likely mean either misusing
+Serde's `Deserialize` trait, or not using Serde at all.
+
+ - [ ] **TODO** discuss/benchmark potential solutions.
+
+Once an event is validated at the source, a new piece of metadata is added to
+the event, indicating against which schema ID the event was validated. This
+metadata can be used by downstream components to perform actions on the event,
+based on the schema. For example, any condition-based transform can use the
+schema ID to pass or fail the condition (e.g. schema-based routing).
+
+### Schema Conversion (TODO)
```

Comment:
I am wondering what this means exactly. It seems like it could be something we leave for later. I like what you've outlined above as an initial step.

Reply:
> I am wondering what this means exactly.

In what sense?

To give a bit more detail (which I'll add to the docs as well); currently we already do "schema conversions" by shaping certain known event fields to match whatever the system at the other end of a sink expects. For example, if a system expects timestamps in a "ts" field using a specific format, we do that conversion implicitly. We do this using a list of hard-coded fields that we know often are required by sinks, and require different formats.

With schemas, and schema conversions, this would no longer have to be special-cased, instead we explicitly define schema conversions from one schema to another, and then do the conversion within the sink. It allows us to add more fields as part of the conversion, be more correct in whatever conversion needs to happen, and stop at boot-time if conversion from schema A to B is expected by the pipeline, but unsupported by the sink.

Reply:
I agree that there's more to work out here, at least from my perspective. Simply having a JSON Schema for an event doesn't provide the same semantic meaning of "this is the timestamp field" as the existing conversions, so I imagine we'd still need some mechanism to identify those well-known fields. Maybe that's what you mean by "explicitly define schema conversions", but it's not totally clear.